src/psyclone/parse/algorithm.py

```diff
@@ -201,14 +236,53 @@ def parse(self, alg_filename):
             # Nothing relevant found.
             raise ParseError(
                 "algorithm.py:parser:parse: Program, module, function or "
-                "subroutine not found in parse tree for file "
-                "'{0}'".format(alg_filename))
+                "subroutine not found in fparser2 parse tree.")
 
         self._unique_invoke_labels = []
         self._arg_name_to_module_name = OrderedDict()
+        self._arg_type_defns = dict()
         invoke_calls = []
 
         for statement in walk(alg_parse_tree.content):
+            if isinstance(statement,
```

Comment:
Could you re-use the routines in the fparser2 frontend that deal with type parsing? You could then just extract the information from the resulting symbol table. That way we don't have duplicate code that examines the types in an fparser2 parse tree.

Reply:
It's a good idea, and one that I didn't think about. I've taken a look and I think the only practical way would be to process all declarations as otherwise we could get missing symbol errors. I don't want to do this (process all declarations) as I am concerned about robustness seeing as we are parsing the LFRic algorithm layer and have not tried the PSyIR processing on this yet. Therefore I'm going to suggest we continue with the existing selective extraction of information from the fparser2 tree which works on the LFRic algorithm code and wait until the Algorithm layer to PSyIR translation is complete which will result in all of this code being replaced.

Reply:
OK, I'm happy with that argument (you'll be pleased to hear).